### ISSUE TITLE
OpenFAST Simulink compilation in dev branch crashing.

### ISSUE THREAD

**Issue by aidowu** [OpenFAST/openfast#692]
I want to ask if you can give a timeline for when the Simulink OpenFAST version (at least up to the TCF MOD2 merge) in the development branch will be ready? 

We tried compiling currently as is, to get started on our work, pending the time the updates gets completed. While the OpenFAST compiled for the command line works, the Simulink compilation keeps crashing MATLAB.

**Comment by andrew-platt** [comment]
How are you compiling it, which compiler, and which platform?

**Comment by aidowu** [comment]
I am compiling following the steps listed in the attached file which I found on this forum. I am compiling for windows x64. It does compile and I generate the S-Function from the compilation but when I try running it, it crashes MATLAB. I have followed the steps in the attached document to compile the main branch and that runs well.
[OpenFAST_Simulink.pdf](https://github.com/OpenFAST/openfast/files/6192553/OpenFAST_Simulink.pdf)

**Comment by andrew-platt** [comment]
Are there any errors from the `create_MATLAB_SFunc.m` script?

**Comment by aidowu** [comment]
There are no errors. Here is a picture of the directory containing the compiled files.
![image](https://user-images.githubusercontent.com/9618674/112214573-bb445c80-8bf5-11eb-9c74-27329ff4e0e8.png)

**Comment by andrew-platt** [comment]
What version of the Intel Fortran compiler are you using with Visual Studio?

Which Simulink example case are you running?

Also, I'm not entirely sure what you are asking about in terms of timeline.  The TCF MODS2 branch was merged in January, but did not contain any updates for control channels.  If you are asking about the cable control channels, there is a branch in development associated with this pull request: #664.

**Comment by aidowu** [comment]
I am using Intel Parallel Studio X17 and I didn't run any of the Simulink example cases but ran a project case that worked with the main dev Simulink compilation. I will go run an example case to see if I can narrow down where the issue might be.

Looking through the pull request you mentioned above, I'll modify my question by asking if the Simulink version in the development branch has been validated for the cable control updates? That way I can know if the error is from my path or if it might be a bug?

**Comment by andrew-platt** [comment]
Did you update the input file set for your project?  Several input files have changed between the main branch and the development branch.  There had been some error handling issues that were causing MATLAB to crash without telling the user what occurred -- all the known ones have been corrected in the development branch.

The development branch does not currently contain any of the cable control functionality.  We hope to merge that in by the end of the month (#664).

**Comment by aidowu** [comment]
Hi Andrew, thanks for the update on the status of the cable control capability. 
Yes, I did update the input file. I found that deleting the temp file created by Simulink helped the simulation advance but I am now getting an error of "size of NumOutputs is invalid". I have attached a screenshot of this error. Can you help provide some insight on why I might be getting this error?
![image](https://user-images.githubusercontent.com/9618674/112501999-14300400-8d60-11eb-8cd8-da95cf7a52d5.png)

**Comment by andrew-platt** [comment]
Which branch are you using?  

If you are using the branch in PR #664, there are some changes in the Simulink interface from the OpenFAST main and dev branches.

**Comment by aidowu** [comment]
Yes, I am using the branch PR #664. Does that mean that I need to change the open loop block I have to the open loop block in the Simulink example folder? 
What I have done initially was just compile the SFunc using the branch PR #664, modify the input files (ServoDyn) and run the simulation using a previous simulink model that worked with the OpenFAST main branch.

**Comment by andrew-platt** [comment]
Yes. You will need to change your Similink model to match the setup in the examples. There are an additional 43 channels in the new block configuration in 3 groups:

- airfoil command: 3
- Cable DeltaL command: 20
- Cable DeltaLDot command: 20

<img width="844" alt="Screen Shot 2021-03-25 at 10 37 21 AM" src="https://user-images.githubusercontent.com/2745453/112509444-15f4ca00-8d56-11eb-8276-1e18627a6e51.png">

**Comment by aidowu** [comment]
Thanks a lot Andrew. I'll make the modifications and see how it goes.

**Comment by andrew-platt** [comment]
Hi @aidowu,

There might be another issue going on here with the output handling within OpenFAST that is causing a mismatch between the list of output channels passed to Simulink and the size of the array holding the output channel data.  The current suspicion is that one of the modules has a bug with this and is causing the `FAST_Start:size of NumOutputs is invalid.` error.

Can you post your outlist section for the HydroDyn input file?

**Comment by aidowu** [comment]
Hi @andrew-p, I attached the output section of my HydroDyn input file (It's only Wave1Elev). 

I am yet to get the Simulink running. It's currently just hanging and no longer getting to the point of returning the "size of NumOutputs is invalid" error after updating the model and input files.

![image](https://user-images.githubusercontent.com/9618674/112913989-eda7fb00-90c8-11eb-94af-4e36c913b695.png)
latt ,

**Comment by bjonkman** [comment]
One other thought: How many output channels are in your model? You should be able to check how many there are in the OpenFAST summary file that is generated at initialization. That error might be triggered if you've exceeded the max number of outputs allowed in the Simulink interface (4000).

**Comment by aidowu** [comment]
Hi @bjonkman, my model has 465 outputs channels.

**Comment by Andrew-P** [comment]
> Hi @Andrew-P, I attached the output section of my HydroDyn input file (It's only Wave1Elev).

I am not involved in this project.  Please direct comments to @andrew-platt, not @Andrew-P.

**Comment by andrew-platt** [comment]
The issue has been found (see #702).  This will be updated shortly in the dev branch and in my branch (for #664).

**Comment by andrew-platt** [comment]
Corrected in `dev` and in branch for #664.

**Comment by andrew-platt** [comment]
PR #702 only partially fixed things.  Now it seems to hang.  Looking into it.

**Comment by Cookiewyf** [comment]
> ‎您是否更新了项目设置的输入文件？主分支和开发分支之间发生了若干输入文件更改。当输入文件错误时，出现了一些错误处理问题，导致 MATLAB 崩溃 - I 这些问题已在开发分支中更正。‎
> 
> ‎开发分支目前不包含任何电缆控制功能。我们希望在月底之前合并‎[‎（#664‎](https://github.com/OpenFAST/openfast/pull/664)).

hi I met the same question, I use matlab(2020b) and I also tried 2019. they all crashed, and I have the openfast-simulink_x64.dll,also created the mexw64.But when I want to run openloop.m, the matlab crashed.(I download the openfast2.6)
![image](https://user-images.githubusercontent.com/33385966/118929059-990d5780-b976-11eb-9145-bc6cb58fcc49.png)

**Comment by jjonkman** [comment]
@rafmudaf -- Can you take these precompiled versions of OpenFAST v2.6 for MATLAB / Simulink on Windows from Bonnie, along with the precompiled executable for Windows and put them on the main OpenFAST v2.6 release page?  Thanks,

**Comment by rafmudaf** [comment]
Yes. I only see the Simulink files. Where is the precompiled executable?

**Comment by jjonkman** [comment]
Haven't you been generating those with each release (except v2.6)?  Otherwise, Bonnie likely has one as well.

**Comment by rafmudaf** [comment]
I generate those manually but I'm working on moving the process to an automated system, so it is delayed. If @bjonkman has already compiled this, I'll add it to the release.

**Comment by bjonkman** [comment]
@rafmudaf , I just uploaded the additional file. Note that it does not require the Intel redistributables.

**Comment by rafmudaf** [comment]
Thanks! Its all now included in the [OpenFAST v2.6.0 release](https://github.com/bjonkman/openfast/releases/tag/v2.6.0).

**Comment by jjonkman** [comment]
Thanks!
